pkg/gsoc/gsoc.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2024 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gsoc
+
+import (
+	"sync"
+
+	"github.com/ethersphere/bee/v2/pkg/log"
+	"github.com/ethersphere/bee/v2/pkg/soc"
+	"github.com/ethersphere/bee/v2/pkg/swarm"
+)
+
+type Listener interface {
+	Subscribe(address [32]byte, handler handler) (cleanup func())
```

Comment:
I just wanted a define a package level struct for handling. why is it not okay? I don't want to export it

Reply:
Interface is defined in order to define a behaviour that can have multiple implementations. I see no reason to have an interface exposed or even created if there is only one implementation and that only the containing packages can implement more. What is the purpose of the interface here if only one implementation exists and no other package is able to create another implementation?

Reply:
`handler` is used at multiple places and the code looks better with that than using `func([]byte)` definitions everywhere IMHO.

Reply:
I wouldn't pollute the code with making a type public when it is not used outside of the package.
`Subscribe` is called in the code and the handler function is not needed to be defined outside of the function call.

if it happens in the future, it may worth to make it public.

Reply:
It would be good to have a comment why the interface is deliberately left unimplementable outside of the package, if you insist in making it as it is.

Reply:
I added a comment line about it.

Reply:
This is not a comment about the interface. Would you, please, add a comment why the interface is made unimplementable by design.

Reply:
because of that I wrote in the comment

Reply:
Janos is pointing out that since the `handler` type is unexported, the `Listener` interface cannot be implemented outside the package. Interfaces are generally intended for multiple implementations. If there will only be a single implementation of Listener (currently within the gsoc package), then it would be more appropriate to use a struct instead of an interface.

The comment is about the `Listener` interface. Not the `handler` type.

Reply:
@nugaon the comment you added is on the handler type https://github.com/ethersphere/bee/pull/4727/commits/691a43d51d2e2ea591543c6fc0620f534bdcb189 does not address any issues related to the interface in this discussion.

And I support to use a struct as per @acha-bill explanation.